### DISCUSSION TITLE
Interesting new upscaler/rotation algo (clear4x)

### DISCUSSION THREAD

**Discussion by JavaGamer** [Orama-Interactive/Pixelorama#793] [Ideas]
I found this new upscaler/rotation algorithm called clear4x on twitter that claims to be better than rotsprite for solid lines and edges:
https://twitter.com/torcado/status/1603530954038992896
https://cohost.org/torcado/post/621636-clean4x-a-pixel-art

It currently implemented as a godot shader so it should be easier[_citation needed_] to be used here?
![](https://staging.cohostcdn.org/attachment/4b46554b-04b7-467b-b685-9249ab0e020c/demo1.png)

**Comment by OverloadedOrama** [comment]
Currently looking into this. The shader itself is easy to implement and works on GLES2, but it doesn't seem to have any rotating logic in it, only scaling. I'm assuming the author is applying the shader on a viewport, and then rotates the images using Godot's regular node rotation. This won't work for Pixelorama, as we're directly manipulating the image data. I'm trying to combine it with nearest neighbour rotation inside the shader, but the result doesn't seem to be the same. I guess we need a way to apply rotation _after_ the image has been scaled, inside the same shader.

We could also use this simply for scaling, along with OmniScale (#511), but I'm also unsure how to scale up the image directly. We need to scale up the image data and not just make them render bigger.

**Comment by OverloadedOrama** [comment]
I opened a pull request #794 if anyone wants to test my current attempt, and/or has any ideas of a better way to handle this.

**Comment by OverloadedOrama** [comment]
Implemented in #794, both for rotation and scaling. It is available in the (unstable) [nightly builds](https://www.oramainteractive.com/Pixelorama-Docs/user_manual/installation#early-access---nightly-versions), and it will be available in the next stable version.
